src/qt3utils/datagenerators/princeton.py

```diff
@@ -0,0 +1,344 @@
+        return value
+
+    def load_experiment(self, value: str) -> None:
+        self.experiment.Load(value)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachDate, False)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachTime, False)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationAttachIncrement, True)
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationIncrementNumber, 0)
+
+    def file_setup(self) -> None:
+        self.set(lf.AddIns.ExperimentSettings.FileNameGenerationBaseFileName, str(uuid.uuid4()))
+    
+    #TODO: Need to actually implement the three methods below so that you can:
+    #      - save a new experiment
+    #      - save what you are currently working on
+    #      - stop the scan 
+        
+    def save_current_experiment(self) -> None:
+        self.experiment.Save()
+    
+    def save_new_experient(self, value) -> None:
+        self.experiment.SaveAs(value)
+
+    def stop_scan(self) -> None:
+        self.experiment.Stop()
+
+    #NOTE: Thw while loop in here will not be needed once you fix the "FileNameGeneration" problem.
+    def start_acquisition_and_wait(self) -> None:
+        """
+        Starts the acquisition process and waits until it is completed before carrying on.
+        """
+        acquisition_time = self.get(lf.AddIns.CameraSettings.ShutterTimingExposureTime)
+        num_frames = self.get(lf.AddIns.ExperimentSettings.AcquisitionFramesToStore)
+        self.experiment.Acquire()
+        
+        sleep(0.001 * acquisition_time * num_frames)  # waiting for the exposure duration
+        
+        while self.experiment.IsRunning:
+            sleep(0.1)  # checking if the experiment is still running
+
+    def process_acquired_data(self) -> Union[np.ndarray, Any]:
+        """
+        Processes the most recently acquired data and returns it as a numpy array.
+        """
+        last_file = self._application.FileManager.GetRecentlyAcquiredFileNames()[0]
+        curr_image = self._application.FileManager.OpenFile(last_file, FileAccess.Read)
+
+        if curr_image.Regions.Length == 1:
+            if curr_image.Frames == 1:
+                return self._process_single_frame(curr_image)
+            else:
+                return self._process_multiple_frames(curr_image)
+        else:
+            raise Exception('curr_image.Regions is not valid. Please retry.')
+
+    def _process_single_frame(self, curr_image: Any) -> np.ndarray:
+        """
+            **Single Frame**:
+            - Data is returned as a 2D array representing raw counts from each pixel.
+            - A vertical section (spanning 400 pixels) represents a range for wavelength averaging.
+        """
+        frame = curr_image.GetFrame(0, 0)
+        return np.reshape(np.fromiter(frame.GetData(), dtype='uint16'), [frame.Width, frame.Height], order='F')
+
+    def _process_multiple_frames(self, curr_image: Any) -> np.ndarray:
```

Comment:
Hmm... what type is `curr_image`?

Reply:
It is of type `IImageDataSet` in C#. What would be the best way to handle this in Python type hints